pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,283 @@
+		// log if there is an issue updating the ingressConfig resource
+		if updateIngressConfigErr := c.updateIngressConfigStatus(ctx, ingressConfigCopy, errors); updateIngressConfigErr != nil {
+			klog.Infof("Error updating ingress with custom route status: %v", err)
+		}
+		return fmt.Errorf("custom route configuration failed verification: %v", errors)
+	}
+
+	// create or modify the existing route
+	if _, _, err = resourceapply.ApplyRoute(ctx, c.routeClient, syncCtx.Recorder(), expectedRoute); err != nil {
+		return err
+	}
+
+	// update ingressConfig status
+	if err = c.updateIngressConfigStatus(ctx, ingressConfigCopy, nil); err != nil {
+		return err
+	}
+
+	// sync the secret
+	return c.syncSecret(secretName)
+}
+
+func (c *customRouteController) getOAuthRouteAndSecretName(ingressConfig *configv1.Ingress) (*routev1.Route, string, []error) {
+	route := c.targetRoute
+	// set defaults
+	route.Spec.Host = route.ObjectMeta.Name + "." + ingressConfig.Spec.Domain // mimic the behavior of subdomain
+	secretName := ""
+
+	// check if a user is overriding route defaults
+	if componentRoute := routecomponenthelpers.GetComponentRouteSpec(ingressConfig, route.ObjectMeta.Namespace, route.ObjectMeta.Name); componentRoute != nil {
+		var errors []error
+		// Check if the provided secret is valid
+		secretName = componentRoute.ServingCertKeyPairSecret.Name
+		if err := c.validateCustomTLSSecret(secretName); err != nil {
+			errors = append(errors, err)
+		}
+
+		// Check if the provided hostname is valid
+		hostname := string(componentRoute.Hostname)
+		if _, err := url.Parse(hostname); err != nil {
+			errors = append(errors, err)
+		}
+
+		if errors != nil {
+			return nil, "", errors
+		}
+
+		route.Spec.Host = hostname
+	}
+
+	return route, secretName, nil
+}
+
+func (c *customRouteController) validateCustomTLSSecret(secretName string) error {
+	if secretName != "" {
+		secret, err := c.secretLister.Secrets("openshift-config").Get(secretName)
+		if err != nil {
+			return err
+		}
+
+		var errors []error
+		privateKeyData, ok := secret.Data[corev1.TLSPrivateKeyKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSPrivateKeyKey))
+		} else {
+			errors = append(errors, certs.ValidatePrivateKey(privateKeyData)...)
+		}
+
+		certData, ok := secret.Data[corev1.TLSCertKey]
+		if !ok {
+			errors = append(errors, fmt.Errorf("custom route secret must include key %s", corev1.TLSCertKey))
+		} else {
+			errors = append(errors, certs.ValidateServerCert(certData)...)
+		}
+
+		if len(errors) != 0 {
+			return fmt.Errorf("error validating secret %s/%s: %v", "openshift-config", secretName, errors)
+		}
+	}
+	return nil
+}
+
+func (c *customRouteController) updateIngressConfigStatus(ctx context.Context, ingressConfig *configv1.Ingress, customRouteErrors []error) error {
+	// update ingressConfig status
+	// TODO: consider using c.targetRoute. Nevertheless the downside is that we might have a stale reference?
```

Comment:
the FLP me wants `targetRoute` to be an argument of this function

Reply:
I checked this piece of code again and the comment is wrong. We can't use the `c.targetRoute` as we extract the `route.Spec.Host`. This can be set either by a user or automatically by the Route Controller. We always need to grab a fresh reference here...

Erasing the comment.